python-client/src/gabriel_client/zeromq_client.py

```diff
@@ -0,0 +1,266 @@
+            source_name (str): The name of the source to produce inputs for
+        """
+        await self._welcome_event.wait()
+
+        source = self._sources.get(source_name)
+        assert source is not None, (
+            "No engines consume frames from source: {}".format(source_name))
+
+        # Async task used to producer an input
+        producer_task = None
+
+        while self._running:
+            try:
+                # Wait for a token. Time out to send a heartbeat to the server.
+                await asyncio.wait_for(source.get_token(), timeout=HEARTBEAT_INTERVAL)
+            except TimeoutError:
+                self._schedule_heartbeat.set()
+                continue
+            # Stop producing inputs if disconnected from the server
+            if not self._connected.is_set():
+                logger.debug("Stopping producer task")
+                if producer_task is not None:
+                    producer_task.cancel()
+                    producer_task = None
+                await self._connected.wait()
+                logger.debug("Resuming producer task")
+
+            if producer_task is None:
+                logger.debug("Created new producer task")
+                producer_task = asyncio.create_task(producer())
+            try:
+                # Wait for the producer to produce an input. Time out to send
+                # a heartbeat to the server. Use an asyncio.shield to prevent
+                # task cancellation.
+                input_frame = await asyncio.wait_for(
+                    asyncio.shield(producer_task), timeout=HEARTBEAT_INTERVAL)
+            except TimeoutError:
+                source.return_token()
+                self._schedule_heartbeat.set()
+                await asyncio.sleep(0)
+                continue
+
+            logger.debug("Got input from producer")
+            producer_task = None
+
+            if input_frame is None:
+                source.return_token()
+                logger.info('Received None from producer')
+                continue
+
+            from_client = gabriel_pb2.FromClient()
+            from_client.frame_id = source.get_frame_id()
+            from_client.source_name = source_name
+            from_client.input_frame.CopyFrom(input_frame)
+
+            # Send input to server
+            await socket.send(from_client.SerializeToString())
+
+            logger.debug('num_tokens for %s is now %d', source_name,
+                         source.get_num_tokens())
+            source.next_frame()
+
+    async def _send_heartbeat(self):
+        """
+        Send a heartbeat to the server.
+        """
+        # Do not send a heartbeat if disconnected from server or a heartbeat
+        # is pending.
+        if not self._connected.is_set() or self._pending_heartbeat:
+            return
+        logger.debug("Sending heartbeat to server")
+        self._pending_heartbeat = True
```

Comment:
Good point, the ZeroMQ tutorial [here](https://zguide.zeromq.org/docs/chapter4/) suggests trying to reconnect the socket if a reply isn't received for a while. I think by default ZeroMQ offers exactly once semantics. But if the server crashes or the reply is lost the client just hangs.

I have revamped the heartbeat logic to periodically reconnect and send a heartbeat if a response isn't received after a timeout. The only thing I skipped is giving up after a few retries instead of trying to reconnect indefinitely.